chain-core/src/common/merkle_tree.rs

```diff
@@ -0,0 +1,643 @@
+use std::vec::IntoIter;
+
+use blake2::Blake2s;
+
+use super::{hash256, H256, H512, HASH_SIZE_256};
+
+// hash(&vec![])
+const EMPTY_HASH: H256 = [
+    105, 33, 122, 48, 121, 144, 128, 148, 225, 17, 33, 208, 66, 53, 74, 124, 31, 85, 182, 72, 44,
+    161, 165, 30, 27, 37, 13, 253, 30, 208, 238, 249,
+];
+
+/// Represents inner tree structure of Merkle Tree
+#[derive(Debug)]
```

Comment:
some binary serialization / encode would be needed

Reply:
This'll part of future PRs (when I start integrating this merkle tree with existing usages).